File: gamelogic/src/main/java/com/alta_v2/game/gamelogic/stage/event/ChangeStageEvent.java
package com.alta_v2.game.gamelogic.stage.event;

public interface ChangeStageEvent {

    static <T extends ChangeStageEvent> T resolve(ChangeStageEvent instance, Class<T> type) {
        if (type.isInstance(instance)) {
            return type.cast(instance);
        }

        throw new ClassCastException("Failed to cast event. Expected: " + type + ", actual: " + instance.getClass());
    }

}

File: gamelogic/src/main/kotlin/com/alta_v2/game/gamelogic/stage/event/ChangeStageEvent.kt
package com.alta_v2.game.gamelogic.stage.event

import com.alta_v2.model.MenuDefinitionModel
import com.alta_v2.model.TiledMapDefinitionModel

interface ChangeStageEvent

data class ChangeMapStageEvent(val mapDefinition: MenuDefinitionModel = MenuDefinitionModel()) : ChangeStageEvent
data class ChangeMenuStageEvent(val mapDefinition: TiledMapDefinitionModel) : ChangeStageEvent